third_party/xla_client/pjrt_computation_client.cc

```diff
@@ -145,26 +175,49 @@ std::vector<ComputationClient::ComputationPtr> PjRtComputationClient::Compile(
-        instance.computation.GetProgramShape().ValueOrDie();
     xla::CompileOptions compile_options;
-    xla::DeviceAssignment device_assignment(client_->device_count(), 1);
-    device_assignment.FillIota(0);
-    compile_options.executable_build_options.set_device_assignment(
-        device_assignment);
-    // TODO(wcromar): set compile_options.argument_layouts, enable strict shapes
-    compile_options.executable_build_options.set_num_partitions(1);
-    compile_options.executable_build_options.set_num_replicas(
-        client_->device_count());
-    compile_options.parameter_is_tupled_arguments =
-        instance.parameter_is_tupled_arguments;
+    if (instance.is_sharded) {
+      // TODO(yeounoh) multi-host, multi-slice configurations
+      compile_options.executable_build_options.set_use_spmd_partitioning(true);
+      compile_options.executable_build_options.set_num_partitions(
+          client_->device_count());
+      compile_options.executable_build_options.set_num_replicas(1);
+      compile_options.parameter_is_tupled_arguments =
+          instance.parameter_is_tupled_arguments;
+
+      // TODO(244391366) verify this is correct for the collectives ops
+      xla::DeviceAssignment device_assignment(1, client_->device_count());
+      device_assignment.FillIota(0);
+      compile_options.executable_build_options.set_device_assignment(
+          device_assignment);
+    } else {
+      // TODO(wcromar): set compile_options.argument_layouts, enable strict
+      // shapes
+      compile_options.executable_build_options.set_num_partitions(1);
+      compile_options.executable_build_options.set_num_replicas(
+          client_->device_count());
+      compile_options.parameter_is_tupled_arguments =
+          instance.parameter_is_tupled_arguments;
+
+      xla::DeviceAssignment device_assignment(client_->device_count(), 1);
+      device_assignment.FillIota(0);
+      compile_options.executable_build_options.set_device_assignment(
+          device_assignment);
+    }
+
+    PjRtDevice* pjrt_device = StringToPjRtDevice(instance.compilation_device);
     std::unique_ptr<xla::PjRtExecutable> executable =
-        client_->Compile(instance.computation, compile_options).ValueOrDie();
+        ConsumeValue(client_->Compile(instance.computation, compile_options));
+
+    const auto& hlo_modules = ConsumeValue(executable->GetHloModules());
+    HloComputation* hlo_computation = hlo_modules[0]->entry_computation();
+    xla::ProgramShape program_shape =
+        xla::ProgramShape(hlo_computation->ToProto().program_shape());
+
     std::shared_ptr<PjRtComputation> pjrt_computation =
-        std::make_shared<PjRtComputation>(std::move(instance.computation),
-                                          program_shape, instance.devices,
-                                          std::move(executable));
+        std::make_shared<PjRtComputation>(
+            std::move(xla::XlaComputation(hlo_modules[0]->ToProto())),
```

Comment:
How is this new `XlaComputation` different than `instance.computation`? Does compilation change the program shape?

Reply:
Yes, we should take the program_shape after the compilation -- it worked before because it didn't change before and after, but it can change depending on the compilation options, including sharding.